pkg/chunk/delete_requests_store.go

```diff
@@ -0,0 +1,253 @@
+package chunk
+
+import (
+	"context"
+	"encoding/binary"
+	"encoding/hex"
+	"errors"
+	"flag"
+	"fmt"
+	"hash/fnv"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+)
+
+type DeleteRequestStatus string
+
+const (
+	Received     DeleteRequestStatus = "0"
+	BuildingPlan DeleteRequestStatus = "1"
+	Deleting     DeleteRequestStatus = "2"
+	Processed    DeleteRequestStatus = "3"
+)
+
+var pendingDeleteRequestStatuses = []DeleteRequestStatus{Received, BuildingPlan, Deleting}
+
+// DeleteRequest holds all the details about a delete request
+type DeleteRequest struct {
+	RequestID string              `json:"request_id"`
+	UserID    string              `json:"-"`
+	StartTime model.Time          `json:"start_time"`
+	EndTime   model.Time          `json:"end_time"`
+	Selectors []string            `json:"selectors"`
+	Status    DeleteRequestStatus `json:"status"`
+	Matchers  [][]*labels.Matcher `json:"-"`
+	CreatedAt model.Time          `json:"created_at"`
+}
+
+// DeleteStore provides all the methods required to manage lifecycle of delete request and things related to it
+type DeleteStore struct {
+	cfg         DeleteStoreConfig
+	indexClient IndexClient
+}
+
+// DeleteStoreConfig holds configuration for delete store
+type DeleteStoreConfig struct {
+	Store             string `yaml:"store"`
+	RequestsTableName string `yaml:"requests_table_name"`
+}
+
+// RegisterFlags adds the flags required to configure this flag set.
+func (cfg *DeleteStoreConfig) RegisterFlags(f *flag.FlagSet) {
+	f.StringVar(&cfg.Store, "deletes.store", "", "Store for keeping delete request")
+	f.StringVar(&cfg.RequestsTableName, "deletes.requests-table-name", "delete_requests", "Name of the table which stores delete requests")
+}
+
+// NewDeleteStore creates a store for managing delete requests
+func NewDeleteStore(cfg DeleteStoreConfig, indexClient IndexClient) (*DeleteStore, error) {
+	ds := DeleteStore{
+		cfg:         cfg,
+		indexClient: indexClient,
+	}
+
+	return &ds, nil
+}
+
+// Add creates entries for a new delete request
+func (ds *DeleteStore) AddDeleteRequest(ctx context.Context, userID string, startTime, endTime model.Time, selectors []string) error {
```

Comment:
A delete request is identified by its ID and userID. If a collision happens between IDs of requests from same user then one of both delete requests would have problems running. I do not think we should throw an error. At best I can check for a collision first and then try recreating the ID since it considers current time as well for the generation of the ID. Any other better way you can think of?

Reply:
Proposed solution sounds good to me. It would be nice if we could detect collision while *writing*, and not as a step before, but hopefully collisions will be rare.

Reply:
With `while writing` I guess you mean db returning an error on collision. I think we can't do that because of how our index stores work. All of them would just update the value instead of throwing an error. Closest is us checking it first with a query.

Reply:
> With `while writing` I guess you mean db returning an error on collision.

Yes.

> I think we can't do that because of how our index stores work. All of them would just update the value instead of throwing an error. Closest is us checking it first with a query.

I assumed this is unfortunately the case :-(